File: content-solutions/queue-deque/deque-stepping-score.py
from collections import deque


def solution(scores, k):
    n = len(scores)
    best = [0] * n
    best[0] = scores[0]
    window = deque([0])
    for i in range(1, n):
        while window[0] < i - k:
            window.popleft()
        best[i] = scores[i] + best[window[0]]
        while window and best[window[-1]] <= best[i]:
            window.pop()
        window.append(i)
    return best[-1]

File: content-solutions/queue-deque/deque-stepping-score.js
function solution(scores, k) {
  const n = scores.length;
  const best = new Array(n).fill(0);
  best[0] = scores[0];
  const window = [0];
  let head = 0;
  for (let i = 1; i < n; i++) {
    while (window[head] < i - k) head++;
    best[i] = scores[i] + best[window[head]];
    while (window.length > head && best[window[window.length - 1]] <= best[i]) window.pop();
    window.push(i);
  }
  return best[n - 1];
}
